File: 2022/Programmers/Level3/sheep_and_wolf.py
visited = [0] * (1 << 17)
answer = 0
def dfs(node:int, state:int, info:list, edges:list):
    global answer
    
    if visited[state]: # 이미 방문한 경로인 경우
        return
    
    print(bin(state), node)
    
    visited[state] = True
    
    wolf, num = 0, 0
    
    # 늑대, 양 세기
    for i in range(len(info)):
        if state & 1 << i: # 방문한 노드라면
            num += 1
            wolf += info[i]
            
    if wolf * 2 >= num:
        return
    
    # 최대 양의 개수 구하기
    answer = max(answer, num - wolf)
    
    # 다음 노드 탐색
    for edge in edges:
        p, c = edge[0], edge[1]
        if p != node:
            continue
        
        if not (state & 1 << c):
            dfs(c, state | 1 << c, info, edges)
        

def solution(info, edges):
    
    dfs(0, 1, info, edges)
    return answer

if __name__ == "__main__":

    info = [0,0,1,1,1,0,1,0,1,0,1,1]
    edges = [[0,1],[1,2],[1,4],[0,8],[8,7],[9,10],[9,11],[4,3],[6,5],[4,6],[8,9]]
    solution(info, edges)

File: 2022/Programmers/Level3/sheep_and_wolf.cpp
#include <bits/stdc++.h>

using namespace std;
int visited[1 << 17]; // 비트마스킹,상태x를 방문했는가?
int l[20], r[20], val[20];
int n;
int answer;

void dfs(int state)
{
    // 이미 방문한 경로인 경우
    if (visited[state])
        return;

    visited[state] = 1;

    int wolf = 0, num = 0;

    for (int i = 0; i < n; i++)
    {
        if (state & (1 << i)) // 방문하지 않은 경로인 경우
        {
            num++;
            wolf += val[i];
        }
    }

    if (2 * wolf >= num)
        return;

    answer = max(answer, num - wolf);

    for (int i = 0; i < n; i++)
    {
        if (!(state & (1 << i)))
            continue;

        if (l[i] != -1)
            dfs(state | 1 << l[i]);

        if (r[i] != -1)
            dfs(state | 1 << r[i]);
    }
}

int solution(vector<int> info, vector<vector<int>> edges)
{
    n = info.size();
    fill(l, l + n, -1);
    fill(r, r + n, -1);

    for (int i = 0; i < n; i++)
        val[i] = info[i];

    for (int i = 0; i < n - 1; i++)
    {
        int a = edges[i][0];
        int b = edges[i][1];

        if (l[a] == -1)
            l[a] = b;
        else
            r[a] = b;
    }

    dfs(1);

    return answer;
}

int main(void)
{
    vector<int> info = {0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1};
    vector<vector<int>> edges = {{0, 1}, {1, 2}, {1, 4}, {0, 8}, {8, 7}, {9, 10}, {9, 11}, {4, 3}, {6, 5}, {4, 6}, {8, 9}};

    solution(info, edges);
    cout << answer << "\n";
    return 0;
}